lib/re_automata.ml

```diff
@@ -307,25 +307,41 @@ module State = struct
     { idx: idx
     ; category: category
     ; desc: E.t list
-    ; mutable status: status option
+    ; status: status Lazy.t
     ; hash: hash }
 
+  let flatten_match m =
+    let ma = List.fold_left (fun ma (i, _) -> max ma i) (-1) m in
+    let res = Array.make (ma + 1) (-1) in
+    List.iter (fun (i, v) -> res.(i) <- v) m;
+    res
+
+  let mk_status = function
+    | []              -> Failed
+    | E.TMatch m :: _ -> Match (flatten_match m.Marks.marks, m.Marks.pmarks)
+    | _               -> Running
+
   let dummy =
     { idx = -1
     ; category = -1
     ; desc = []
-    ; status = None
+    ; status = lazy (mk_status [])
     ; hash = -1 }
 
   let hash idx cat desc =
     E.hash desc (hash_combine idx (hash_combine cat 0)) land 0x3FFFFFFF
 
-  let mk idx cat desc = 
-    { idx
-    ; category = cat
-    ; desc
-    ; status = None
-    ; hash = hash idx cat desc}
+  let mk idx category desc =
+    let rec t =
+      { idx
+      ; category
+      ; desc
+      ; status
+      ; hash = hash idx category desc}
+    and status = lazy (mk_status t.desc) in
```

Comment:
I thought that in my version if the desc is updated before status is forced, we'll be using the updated version of desc to make status. This is too much the semantics of the existing status more closely. I could be wrong though.